Sidebar Footer: Add new, more flexible, layout

jasmussen opened the pull request:
This PR <strike>starts work on</strike> fixes #8461. It:

- Adjusts the layout to fit the mockups
- Adds the "Add new site" button to sit there permanently, so we don't have to have a separate switchin layout for the site selector.

Still lots to do.

- [x] <strike>Tracks for "Add new site" wasn't ported.</strike> — irrelevant now
- [x] Multisite switcher needs to work with this new design.
- [x] Unread dot design for chat button.
- [x] Currently the SVG is included directly, instead of referring to a gridicon. Pending PR: https://github.com/Automattic/wp-calypso/pull/8579

Screenshot:

![screen shot 2016-10-07 at 12 27 54](https://cloud.githubusercontent.com/assets/1204802/19187037/97bb08ba-8c89-11e6-8307-b449010876ec.png)

jasmussen commented:
PR to add the chat icon to Calypso here: https://github.com/Automattic/wp-calypso/pull/8579

jasmussen commented:
Added unread dot that can be invoked by applying `sidebar__footer__support-chat-unread` as a CSS class to the chat button. Looks like this:

![screen shot 2016-10-07 at 14 20 32](https://cloud.githubusercontent.com/assets/1204802/19189654/6faee818-8c99-11e6-94df-927afd7d6aaa.png)

mtias commented on client/layout/sidebar/style.scss:
```diff
@@ -346,9 +348,48 @@ a.sidebar__button {
+
+	}
+
+	&.sidebar__footer__support-chat-unread {
```
I suggest using `has-unread` as the class:

```
.sidebar__footer-chat {
    ...
    &.has-unread {
        ...
    }
}
```

jasmussen commented:
Rebased and addressed your issues, Matias, thanks, as well as some others. Is the linter something I have to add manually? 

Now looks like this:

<img width="328" alt="screen shot 2016-10-10 at 12 00 53" src="https://cloud.githubusercontent.com/assets/1204802/19232706/8046b580-8ee1-11e6-922b-64c2086087e2.png">

I could use some code help with the remaining issues, thanks.

jasmussen commented:
CC: @beaucollins because I noticed you abstracted out the sidebar footer button into `<HappychatButton />`. Also, there's now actually a `chat` gridicon we can use instead of `comment`.

beaucollins commented:
@jasmussen I rebased this branch to take into account #7791.

There is a new warning now that the classname of the button in `components/happychat/button.jsx` should be prefixed with `happychat__`.

I'm not sure where that leaves us with the css situation now that the Happychat specific styles should probably be moved to the happychat component's stylesheet. There may be a [less specific classname we can use here](https://github.com/Automattic/wp-calypso/pull/8578/files#diff-71484b48f8a3de77abd86c655a4702d2R358).

jasmussen commented:
> I'm not sure where that leaves us with the css situation now that the Happychat specific styles should probably be moved to the happychat component's stylesheet. 

I could go either way here. 

Would the happychat component ever be shown in a place that isn't the sidebar footer? 

Should the sidebar footer not have design styles that affect buttons we might add to it there? 

It seems like an argument could be made in favor of making the sidebar-footer styles more generic to encompass any buttons we might put there, and then just making the happychat button almost design-free? @mtias any opinions?

mtias commented:
This seems like sidebar specific styles for me, either under the `.sidebar__footer-chat` or a more generic `.sidebar__footer .button`. The first is probably better for now.

jasmussen commented on client/layout/sidebar/footer.jsx:
```diff
@@ -15,8 +15,11 @@ import HappychatButton from 'components/happychat/button';
 		{ children }
-		<Button compact borderless href="/help">
-			<Gridicon icon="help-outline" /> { translate( 'Help' ) }
+		<Button borderless href={ config( 'signup_url' ) + '?ref=calypso-selector' }>
```
Yes, I would say so. The benefit of this new design is that it can be the same everywhere. Help and Add New Site are both worth it, I would say, and the Chat button has the potential to become that.

Keep in mind that for launch, the Happychat button will *appear* when chat is available, or a chat hasn't ended. So it won't be a permanently visible button always.

jasmussen replied:
I hear you — upon reflection, perhaps "Add New Site" shouldn't be that prominent. Perhaps we should keep the current behavior:

- Add New Site only on the site selector (perhaps redesigned with bigger icon and footer bar to match this new style)
- Add New Site permanently footer visible in My Sites, for single-site users. 

However I disagree that "Add New Post" should be added to the footer. Doing so is essentially admitting defeat with this button:

![screen shot 2016-10-19 at 13 11 31](https://cloud.githubusercontent.com/assets/1204802/19516712/93e416fa-95fd-11e6-939f-b7aa53a1d2f1.png)

If that button doesn't work as intended, we could consider beefing it up at desktop breakpoints:

![screen shot 2016-10-19 at 13 12 49](https://cloud.githubusercontent.com/assets/1204802/19516757/cf119176-95fd-11e6-9641-77592e0c1240.png)

But having multiple buttons that do the same almost always makes the user pose the question, "which one is the right one?"

mtias replied:
Oh, I think I didn't explain myself correctly — the sidebar footer area is setup in a way that any section can provide their custom functionality, while we keep some thing consistent throughout (help and chat buttons).

My idea was that when you are IN the editor we could show an add new post button, not in my sites. Starting a new post in the current site is not as straightforward as it should; the masterbar new post item forces you to reselect a site. Since "Add new site" makes sense for a footer action in `My Sites`, I was saying that "Add new post" may make sense as the primary footer action of the `Editor`.

jasmussen replied:
Ah! Understood. That functionality should still technically be there:

```
	<div className="sidebar__footer">
		{ children }
```

Incidentally removing the hard-coded Add New Site button from the footer fixes Lance's initial issue.

rickybanister replied:
I think that would be very confusing actually if that sidebar action changed based on location.

folletto commented:
> I'm not sure "add new site" warrants such a prominent display across all sections. For instance, in the editor I think we could have an "add new post" instead, that is more relevant and fills a void we have at the moment (there's no one-click way to start a new post from the editor in the current site).

After our previous discussion, and seeing the implication of this live, I think I'm leaning in agreeing with this. It feels out of place to have "Add new site" everywhere.

Revisiting old ideas, what about:

In My Sites showing:
```
| Add New site            | (help icon) | (chat icon) |
```

Everywhere else  
```
|                         | (help icon) | (chat icon) |
```

In the Editor
```
| Add new draft           | (help icon) | (chat icon) |
```

Does it actually make sense at all to have the footer in the Editor?

The blank space might be slightly off, but I think it still works well as a whole, without shifting items around that I feel would be even worse.

jasmussen commented:
> The blank space might be slightly off, but I think it still works well as a whole, without shifting items around that I feel would be even worse.

If we _do_ decide that we want to shift the contents of the left side around, I feel like at the least we should keep the footer around all the time. Help is always useful, and the chat might be in the future too. Keep in mind it can hold notifications too. 

What goes in that area to the left is open to discussion, I don't feel too strongly about it, but the whitespace doesn't worry me — there's plenty of whitespace above it most of the time. 

I'm working on something right now where we can get a video test of it. Stay tuned.

folletto commented:
> What goes in that area to the left is open to discussion, I don't feel too strongly about it, but the whitespace doesn't worry me — there's plenty of whitespace above it most of the time.

Then I'm ok with that. :)

jasmussen commented:
Pushed a change to make the menu behave like this instead:

![oct-19-2016 13-44-56](https://cloud.githubusercontent.com/assets/1204802/19517580/5e1d4780-9602-11e6-999d-bb6e1a47db37.gif)

Incidentally, if we're happy with this, that *should* mean all that's left to do is verify this works with single site users. Any tips on how to do that?

jasmussen commented:
Thanks to @folletto, I managed to test and fix users with only a single blog. It looks like this now:

![screen shot 2016-10-19 at 13 59 14](https://cloud.githubusercontent.com/assets/1204802/19517936/67df4460-9604-11e6-80da-28d9516540b2.png)

Here's a full shot:

![screen shot 2016-10-19 at 13 59 37](https://cloud.githubusercontent.com/assets/1204802/19517939/6d4f8338-9604-11e6-8863-69eed363763a.png)

Incidentally, I think that actually wraps up this PR and makes it ready for final review. Technically all is there now — we just have to agree that this is the footer design and behavior we want. That is:

- New, bigger footer, with permanent help button, and optional chat button
- Footer is present everywhere, except for the "Me" section
- Chat button has optional unread state and new chat gridicon
- When a single-site user visits "My Sites", they see an Add New Site link
- When a multi-site user visits My Sites, they do not see the Add New Site link, they have to click "Switch", where they see the Add New Site link there
- Site selector footer has been updated to the new design

Lay all your thoughts on me.

folletto commented:
> Footer is present everywhere, except for the "Me" section

Why this?

Tested, looks and works great. I think we can ship this and then refine as needed. 👍

jasmussen commented:
> Why this?

Initially when we introduced the footer, it primarily held "Help", which was already accessible from the Me section. I think it was considered redundant. We can revisit this decision, I suppose :)

folletto commented:
> We can revisit this decision, I suppose :)

Ok, no need to revisit in this PR imho. Let's move forward. :)

jasmussen commented:
Got lots of thumbs up for the design so far, removing the "needs design review" tag.

mjangda commented:
👍

rickybanister commented:
I have a couple of questions—I realize the /me menu has a help item, but wouldn't it be easier to find if it was always in the footer across all sections, not jumping around?

And am I correct that we would not show the 'add site' link in the sidebar if you have more than one site? Why did we decide not to just have the footer be the same for all users all the time in all sections? I feel like this redesign made that possible and was excited by the prospect.

jasmussen commented:
@rickybanister 

> I realize the /me menu has a help item, but wouldn't it be easier to find if it was always in the footer across all sections, not jumping around?

I'm okay with showing the footer on all sections. Now that the footer help icon is icon only anyway, I don't think the original reason for not showing the footer there is valid anymore. 

> And am I correct that we would not show the 'add site' link in the sidebar if you have more than one site? Why did we decide not to just have the footer be the same for all users all the time in all sections? I feel like this redesign made that possible and was excited by the prospect.

That is correct with regards to the current state of this PR. When I started work, I was also excited about the prospect of a unified, simplified footer that didn't change on a per-section basis. 

However that aspect received some concern, both <a href="https://github.com/Automattic/wp-calypso/pull/8578#issuecomment-254786421">here</a> and <a href="https://github.com/Automattic/wp-calypso/pull/8578/files/72dec60eb2e9e8ba85513e41ae2055c87c6a4956#r84045941">here</a>.

In the spirit of shipping, I reverted back to the old behavior, but with the new design. That is, the sidebar footer can receive a prop, `{ children }`, which defines what goes on the left. When you only have a single site, the Add New Site button is passed in as a prop. Incidentally that also made this PR a whole lot simpler.

What do you think about me merging in this PR as is, and then continuing the permanent "add new site" button and presence on the Me section separately?

rickybanister commented:
@jasmussen that sounds 100% fair to me

jasmussen commented:
Opened an issue for discussing a persistent sidebar, and presence on Me section in https://github.com/Automattic/wp-calypso/issues/8885
